fix #1937 Account for requests made to upstream in FluxBufferPredicate

simonbasle opened the pull request:
This commit adds additional request accounting to FluxBufferPredicate,
allowing it to keep track of when there is remaining upstream request
and when there is not. In turn, this allows the "just add to buffer"
case to perform an additional `request(1)` ONLY if necessary (to further
fill the current buffer).

supersedes #2027 cc @UgiR @robotmrv

robotmrv commented on reactor-core/src/main/java/reactor/core/publisher/FluxBufferPredicate.java:
```diff
@@ -144,7 +151,8 @@ public void request(long n) {
 					fastpath = true;
-					requested = Long.MAX_VALUE;
+					requestedBuffers = Long.MAX_VALUE;
+					requestedBuffers = Long.MAX_VALUE;
```
two times `requestedBuffers` ?

simonbasle replied:
nice catch 😅

robotmrv commented on reactor-core/src/main/java/reactor/core/publisher/FluxBufferPredicate.java:
```diff
@@ -156,11 +164,11 @@ public void request(long n) {
 							this,
 							this)) {
-						s.request(1);
+						REQUESTED_FROM_SOURCE.addAndGet(this, n);
```
should not it be added by `Operators#addCap()` ?

simonbasle replied:
indeed better to use capped add 👍

robotmrv commented on reactor-core/src/main/java/reactor/core/publisher/FluxBufferPredicate.java:
```diff
@@ -205,25 +213,31 @@ public boolean tryOnNext(T t) {
 			}
 
-			return !requestMore;
+			boolean isNotExpectingFromSource = REQUESTED_FROM_SOURCE.decrementAndGet(this) == 0;
```
it will take a lot of time to go down to zero, but anyway why to decrement for a fast path?

simonbasle replied:
not sure what you mean. maybe that this decrement operation should be skipped in case `fastpath == true` ?

simonbasle replied:
we need to keep separate track of "how many buffers are wanted by downstream" (`REQUESTED_BUFFERS`) vs "how many elements have been requested upstream" (`REQUESTED_FROM_SOURCE`), so that when we're filling a buffer and we see that said buffer is not yet full, we only ask the source for one more element IFF there is no pending request to it already. Otherwise, we know that we should soon receive at least another element which might fulfill the buffer.

robotmrv replied:
> maybe that this decrement operation should be skipped in case fastpath == true ?

yes, I mean this

simonbasle replied:
updated, by extracting `if (fastpath) return true;` we take care of the fastpath case more clearly and avoid these atomic operations altogether

simonbasle commented:
@robotmrv thanks for the useful review(s) 👍 
I have rebased on top of current 3.2.x and squashed the commits into one, ready for merge if everything looks good to you and @bsideup

robotmrv commented on reactor-core/src/main/java/reactor/core/publisher/FluxBufferPredicate.java:
```diff
@@ -156,11 +164,11 @@ public void request(long n) {
 							this)) {
-						s.request(1);
+						Operators.addCap(REQUESTED_FROM_SOURCE, this, n);
+						s.request(n);
```
@simonbasle I have one question about that you said before  
https://github.com/reactor/reactor-core/pull/2027#discussion_r374182115
> a Subscriber is responsible for ensuring the upstream requests it makes are not done in parallel (rule 2.7) so that line MUST NOT be executed at the same time

as far as I understand `s.request()` could be invoked from `BufferPredicateSubscriber#request()` and from `BufferPredicateSubscriber#onNext()` and they could be invoked by different `Thread`s.  
 How does `rule 2.7` is satisfied here or it is not applicable here?

simonbasle replied:
mmh with the `REQUESTED_FROM_SOURCE` guard, but we indeed don't really use that to ensure `n` is correct here. let's unpack that a little:
 - `addCap` should lose to `compareAndSet` (other writes "happens before")
 - since parallel requests are illegal, the only way addCap loses is if `compareAndSet` switched from 0 to 1, so at that point `requestedFromSource` == `0` or `1`
 - it should thus be safe to use the return value of `addCap` (0 or 1) as a modifier to `n` for the parent request:

```java
long delta = Operators.addCap(REQUESTED_FROM_SOURCE, this, n);
s.request(n - delta);
```

simonbasle replied:
the above doesn't seem to be necessary. I have added tests around racing `request` and `onNext` and validated the current code doesn't lead to backpressure errors.

simonbasle commented:
@reactorbot fast-path

simonbasle commented:
@reactorbot fast path
